iceoryx_utils/include/iceoryx_utils/internal/cxx/storable_function.hpp

```diff
@@ -0,0 +1,162 @@
+// Copyright (c) 2020 by Apex.AI Inc. All rights reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef IOX_UTILS_STORABLE_FUNCTION_HPP
+#define IOX_UTILS_STORABLE_FUNCTION_HPP
+
+#include "iceoryx_utils/cxx/function_ref.hpp"
+#include "iceoryx_utils/cxx/type_traits.hpp"
+#include "iceoryx_utils/internal/cxx/static_storage.hpp"
+
+#include <iostream>
+#include <type_traits>
+#include <utility>
+
+namespace iox
+{
+namespace cxx
+{
+template <typename ReturnType, typename... Args>
+using signature = ReturnType(Args...);
+
+template <typename StorageType, typename T>
+class storable_function;
+
+template <typename StorageType, typename ReturnType, typename... Args>
+class storable_function<StorageType, signature<ReturnType, Args...>>
+{
+  private:
+    // note that this vtable or a similar approach with virtual is needed to ensure we perform the correct
+    // operation with the underlying (erased) type
+    // this means storable_function cannot be used where pointers become invalid, e.g. across process boundaries
+    struct vtable
+    {
+        // exposing those is intentional, it is an internal structure
+        void (*copyFunction)(const storable_function& src, storable_function& dest){nullptr};
+        void (*moveFunction)(storable_function& src, storable_function& dest){nullptr};
+        void (*destroyFunction)(storable_function& f){nullptr};
+
+        vtable() = default;
+        vtable(const vtable& other) = default;
+        vtable& operator=(const vtable& other) = default;
+        vtable(vtable&& other) = default;
+        vtable& operator=(vtable&& other) = default;
+
+        void copy(const storable_function& src, storable_function& dest)
+        {
+            if (copyFunction)
+            {
+                copyFunction(src, dest);
+            }
+        }
+
+        void move(storable_function& src, storable_function& dest)
+        {
+            if (moveFunction)
+            {
+                moveFunction(src, dest);
+            }
+        }
+
+        void destroy(storable_function& f)
+        {
+            if (destroyFunction)
+            {
+                destroyFunction(f);
+            }
+        }
+    };
+
+    vtable m_vtable;
+
+    /// @note  in general we cannot know the alignment of the type we want to store at construction time
+    StorageType m_storage;
```

Comment:
I might be wrong, but I thought the convention was to put member variables at the end of the class definition.

Reply:
Can do. Needed a private section for internal types at the beginning and used it for members. We have no strict convention for this AFAIK. I prefer defining types, then the public API, then the private API (including members and potentially private types).

Reply:
Would be nice to have some kind of convention. I completely missed these members when scanning the file.